File: python/day-10/b.py
from typing import List


def main():
    lines = open('in').read().split()
    incomplete_lines = [line for line in lines if not is_corrupt(line)]
    closing_sequences = [find_closing_sequence(line) for line in lines]
    score = get_score(closing_sequences)
    print(score)


def is_corrupt(line: str) -> bool:
    open_chunk_chars = ['(', '[', '{', '<']
    close_chunk_pairs = {
        ')': '(',
        ']': '[',
        '}': '{',
        '>': '<'
    }

    stack = []
    for c in line:
        if c in open_chunk_chars:
            stack.append(c)
        elif stack.pop() != close_chunk_pairs[c]:
            return True
            
    return False


def find_closing_sequence(line: str) -> str:
    open_chunk_chars = ['(', '[', '{', '<']
    open_chunk_pairs = {
        '(': ')',
        '[': ']',
        '{': '}',
        '<': '>'
    }

    stack = []
    for c in line:
        if c in open_chunk_chars:
            stack.append(c)
        else:
            stack.pop()
            
    return ''.join(reversed([open_chunk_pairs[c] for c in stack]))


def get_score(sequences: List[str]) -> int:
    scores = sorted([get_score_for_sequence(seq) for seq in sequences])
    return scores[(len(scores) - 1) // 2]


def get_score_for_sequence(seq: str) -> int:
    char_scores: dict[str, int] = {
        ')': 1,
        ']': 2,
        '}': 3,
        '>': 4
    }
    score: int = 0
    
    for c in seq:
        score *= 5
        score += char_scores[c]

    return score


main()

File: rust/day-10/b.rs
use std::fs;
use std::error::Error;
use std::collections::HashMap;


fn main() -> Result<(), Box<dyn Error>>{
    let inp = fs::read_to_string("in")?;
    let incomplete_lines: Vec<&str> = inp
        .lines()
        .filter(|line| !is_corrupt(line))
        .collect();
    let closing_sequences: Vec<Vec<char>> = incomplete_lines
        .iter()
        .map(|line| closing_sequence(line))
        .collect();
    let score: usize = get_score(closing_sequences);
    
    println!("{}", score);
    Ok(())
}


fn is_corrupt(line: &str) -> bool{
    let open_chunk_chars: Vec<char> = vec!['(', '[', '{', '<'];

    let mut close_chunk_pairs = HashMap::new();
    close_chunk_pairs.insert(')', '(');
    close_chunk_pairs.insert(']', '[');
    close_chunk_pairs.insert('}', '{');
    close_chunk_pairs.insert('>', '<');

    let mut stack: Vec<char> = Vec::new();
    for c in line.chars() {
        if open_chunk_chars.contains(&c) {
            stack.push(c);
        } else if stack.pop().unwrap() != close_chunk_pairs[&c] {
            return true;
        }
    }
            
    false
}

fn closing_sequence(line: &str) -> Vec<char> {
    let open_chunk_chars: Vec<char> = vec!['(', '[', '{', '<'];

    let mut open_chunk_pairs = HashMap::new();
    open_chunk_pairs.insert('(', ')');
    open_chunk_pairs.insert('[', ']');
    open_chunk_pairs.insert('{', '}');
    open_chunk_pairs.insert('<', '>');

    let mut stack: Vec<char> = Vec::new();
    for c in line.chars() {
        if open_chunk_chars.contains(&c) {
            stack.push(c);
        } else {
            stack.pop();
        }
    }
    
    stack
        .iter()
        .rev()
        .map(|c| open_chunk_pairs[c])
        .collect()
}


fn get_score(sequences: Vec<Vec<char>>) -> usize {
    let mut scores: Vec<usize> = sequences
        .iter()
        .map(|seq| get_score_for_sequence(seq))
        .collect();
    scores.sort();
    scores[(scores.len() - 1) / 2]
}

fn get_score_for_sequence(chars: &Vec<char>) -> usize {
    let mut char_scores = HashMap::new();
    char_scores.insert(')', 1);
    char_scores.insert(']', 2);
    char_scores.insert('}', 3);
    char_scores.insert('>', 4);

    let mut score = 0;
    for c in chars {
        score *= 5;
        score += char_scores[&c];
    }

    score
}